[Federation][e2e] Add 2 new testcases to federation service e2e

shashidharatd opened the pull request:
**What this PR does / why we need it**:
Add 2 new test cases for federation services
- Federation service updation should update clustered service
- Federation service controller should recreate service shard in cluster, if it gets deleted.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #27623, #35827
Handles one of the tasks discussed in #41253

**Special notes for your reviewer**:

**Release note**:
`NONE`

cc @kubernetes/sig-federation-bugs, @nikhiljindal @madhusudancs

Commit:
test service shard recreation when deleted from federated cluster

Commit:
Add Federated Service update test case

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41254)
<!-- Reviewable:end -->

madhusudancs commented on test/e2e_federation/federated-service.go:
```diff
@@ -92,15 +92,15 @@ var _ = framework.KubeDescribe("Federated Services [Feature:Federation]", func()
 		})
 
-		Describe("service creation", func() {
+		Describe("Federated Service", func() {
```
@shashidharatd is there a reason why you think any of these issues are causing resource leaks?

shashidharatd replied:
cleanupServiceShardsAndProviderResources is using `nsName` variable to cleanup the namespace. But this variable is redefined in some testcases like in [here](https://github.com/kubernetes/kubernetes/pull/41254/commits/15fa5947f0033c684106e9ba30dfb84ad6a546e2#diff-cccd4bd82f7a2fff882f5ac20abb960fL132). so actually after test the nsName variable is "" because we never initialized in first place.

madhusudancs replied:
@shashidharatd Wow! Excellent catch! Is this PR independent of your service controller refactor? Can we get this in ASAP?

shashidharatd replied:
yes @madhusudancs, this PR is independent. However the 2 new testcase on services will fail with the current code.

shashidharatd commented:
yes @madhusudancs, this PR is independent. However the 2 new testcase on
services will fail with the current code.

On Tue, Feb 14, 2017 at 11:35 AM, Madhusudan.C.S <notifications@github.com>
wrote:

> *@madhusudancs* commented on this pull request.
> ------------------------------
>
> In test/e2e_federation/federated-service.go
> <https://github.com/kubernetes/kubernetes/pull/41254>:
>
> > @@ -92,15 +92,15 @@ var _ = framework.KubeDescribe("Federated Services [Feature:Federation]", func()
>  			clusters, primaryClusterName = getRegisteredClusters(UserAgentName, f)
>  		})
>
> -		Describe("service creation", func() {
> +		Describe("Federated Service", func() {
>
> @shashidharatd <https://github.com/shashidharatd> Wow! Excellent catch!
> Is this PR independent of your service controller refactor? Can we get this
> in ASAP?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41254>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIVItBff7kBPZ_86kbHv3_x1zqSHYRc5ks5rcUQMgaJpZM4L9qwi>
> .
>

madhusudancs commented:
@shashidharatd Ok. Can you separate them out? I want to merge the nsName fix ASAP.

shashidharatd commented:
ok @madhusudancs, i will separate and raise a PR soon.

shashidharatd commented:
@k8s-bot gci gce e2e test this
@k8s-bot gci gke e2e test this

shashidharatd commented:
@k8s-bot gci gce e2e test this

shashidharatd commented:
@k8s-bot kubemark e2e test this
@k8s-bot verify test this
@k8s-bot cvm gce e2e test this

ethernetdan commented:
Removing issue from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

calebamiles commented:
Added this issue back into the milestone @ethernetdan, only additional test coverage has been added by this PR. Please rebase and get an `lgtm` from your reviewer. Please also watch out for any flakes @shashidharatd.

**cc: @kubernetes/kubernetes-release-managers**

shashidharatd commented:
@k8s-bot gci gke e2e test this

shashidharatd commented:
@k8s-bot gce etcd3 e2e test this

shashidharatd commented:
@madhusudancs, request to move the milestone to 1.7 for this PR

madhusudancs commented:
@shashidharatd done.

shashidharatd commented:
Thanks @madhusudancs

shashidharatd commented:
@k8s-bot kops aws e2e test this

shashidharatd commented:
@madhusudancs, after the PR #41258 to re-factor federation service controller is merged, these two new test cases would pass in the federation e2e tests. Here is the link to test [result](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41254/pull-kubernetes-federation-e2e-gce/2226)
This PR would fix the 2 issues #27623, #35827
request to PTAL and process asap.

madhusudancs commented:
@shashidharatd minor nits.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e_federation/service.go, line 120 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiQ6n4XDnm_lK75lw9:-KiiQ6n4XDnm_lK75lwA:b-84voup) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L120)):*
> ```Go
> 				}()
> 				waitForServiceShardsOrFail(nsName, service, clusters)
> 				By(fmt.Sprintf("Successfuly created and synced service %q/%q to clusters", nsName, service.Name))
> ```

s/Successfuly/Successfully/

I would use a `By()` clause before the operation to indicate that we are starting the operation and if necessary use the logging library to log the successful completion of the operation. `By()` is useful to track the steps and hence is more useful to show that before the operation, in case the operation hangs or fails. And then log to just ensure that there is a trace of its completion.

Also, can you use %q with namespaceName helper function or use `\"%s/%s\"`.? `%q/%q` makes it difficult to search through the logs for `namespace/name`s.

---

*[test/e2e_federation/service.go, line 123 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiT3zYE8N2P1Fqh-AX:-KiiT3zYE8N2P1Fqh-AY:bwmesx4) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L123)):*
> ```Go
> 				service = updateServiceOrFail(f.FederationClientset, nsName, FederatedServiceName)
> 				waitForServiceShardsOrFail(nsName, service, clusters)
> 				By(fmt.Sprintf("Successfuly updated and synced service %q/%q to clusters", nsName, service.Name))
> ```

Same comments as above.

---

*[test/e2e_federation/service.go, line 156 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiSrXbOhcMn7nU0t1r:-KiiSrXbOhcMn7nU0t1s:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L156)):*
> ```Go
> 				}()
> 				waitForServiceShardsOrFail(nsName, service, clusters)
> 				By(fmt.Sprintf("Successfuly created and synced service %q/%q to clusters", nsName, service.Name))
> ```

Same as above.

---

*[test/e2e_federation/service.go, line 164 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiU7GQOP06dvQQKMiu:-KiiU7GQOP06dvQQKMiv:bw64nsu) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L164)):*
> ```Go
> 
> 				waitForServiceShardsOrFail(nsName, service, clusters)
> 				By(fmt.Sprintf("Successfuly recreated service shard %q/%q in %q cluster", nsName, service.Name, primaryClusterName))
> ```

Same as above here and everywhere else.

---

*[test/e2e_federation/service.go, line 340 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiVetkfvRPGZRUuM41:-KiiVetkfvRPGZRUuM42:b-83sjoh) ([raw file](https://github.com/kubernetes/kubernetes/blob/2d79d53fb28cf994f9c0a767f6fd8baa517b611a/test/e2e_federation/service.go#L340)):*
> ```Go
> 	clusterService.Spec.ExternalIPs = federationService.Spec.ExternalIPs
> 	clusterService.Spec.LoadBalancerIP = federationService.Spec.LoadBalancerIP
> 	clusterService.Spec.LoadBalancerSourceRanges = federationService.Spec.LoadBalancerSourceRanges
> ```

Nice!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-:-KiiVlrSl1a9VGVnGTYm:b-xts23i)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[test/e2e_federation/service.go, line 388 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiWBYz-ptBawIdlN9q:-KiiWBYz-ptBawIdlN9r:b-vf2sfx) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L388)):*
> ```Go
> 		return false
> 	}
> 	if !reflect.DeepEqual(federationService.Labels, clusterService.Labels) && (len(federationService.Labels) != 0 || len(clusterService.Labels) != 0) {
> ```

What's the point in this length check? What if someone writes a test with no labels?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, addressed comments. PTAL

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[test/e2e_federation/service.go, line 120 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiQ6n4XDnm_lK75lw9:-KiibfkOqTsm-LZiswi4:b-w7v5ca) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L120)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/Successfuly/Successfully/

I would use a `By()` clause before the operation to indicate that we are starting the operation and if necessary use the logging library to log the successful completion of the operation. `By()` is useful to track the steps and hence is more useful to show that before the operation, in case the operation hangs or fails. And then log to just ensure that there is a trace of its completion.

Also, can you use %q with namespaceName helper function or use `\"%s/%s\"`.? `%q/%q` makes it difficult to search through the logs for `namespace/name`s.
</blockquote></details>

Updated as advised.

---

*[test/e2e_federation/service.go, line 123 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiT3zYE8N2P1Fqh-AX:-Kiic_Ki9TEDfECo_62j:b-w7v5ca) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L123)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Same comments as above.
</blockquote></details>

Updated as advised.

---

*[test/e2e_federation/service.go, line 156 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiSrXbOhcMn7nU0t1r:-KiicahTjalOQyl76GTe:b-w7v5ca) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L156)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Updated as advised.

---

*[test/e2e_federation/service.go, line 164 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiU7GQOP06dvQQKMiu:-Kiicbvye1ZUJIMEdcdq:b-w7v5ca) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L164)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Same as above here and everywhere else.
</blockquote></details>

Updated as advised.

---

*[test/e2e_federation/service.go, line 388 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiWBYz-ptBawIdlN9q:-KiicgGzDUCsw148EoOV:b4h14cq) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L388)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

What's the point in this length check? What if someone writes a test with no labels?
</blockquote></details>

It is to check if labels are also propogated. We shall remove it when we need that labels should not be propagated to service in federated clusters.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-:-KiidD2UoJtMdG07RhCC:b-q8uoct)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@k8s-bot bazel test this
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this

shashidharatd commented:
@madhusudancs, all tests are passing. PTAL

madhusudancs commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e_federation/service.go, line 388 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-KiiWBYz-ptBawIdlN9q:-KiletAZoDDbVW1VlLFr:b-nt9cve) ([raw file](https://github.com/kubernetes/kubernetes/blob/04d941bf5217d3e39ce835cc1767e64635567553/test/e2e_federation/service.go#L388)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

It is to check if labels are also propogated. We shall remove it when we need that labels should not be propagated to service in federated clusters.
</blockquote></details>

I mean, what does `(len(federationService.Labels) != 0 || len(clusterService.Labels) != 0)` check that `reflect.DeepEqual()` doesn't?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
/lgtm
/approve

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41254#-:-KilfAewJATALhqEbcN_:b6yv7yi)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
> I mean, what does (len(federationService.Labels) != 0 || len(clusterService.Labels) != 0) check that reflect.DeepEqual() doesn't?

@madhusudancs, it is to overcome an issue where reflect.DeepEqual treats empty map and nil map as not equal and we would like them to treated as equal. here is the link to the issue. https://github.com/golang/go/issues/16531

madhusudancs commented:
@shashidharatd Ok, makes sense now.
